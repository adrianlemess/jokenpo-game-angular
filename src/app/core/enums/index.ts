export * from './end-game-result';
export * from './hand-type';
export * from './player-type';
