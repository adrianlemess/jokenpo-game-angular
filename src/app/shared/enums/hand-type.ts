export enum HAND_TYPE {
    PAPER = 'paper',
    ROCK = 'rock',
    SCISSOR = 'scissor',
}
