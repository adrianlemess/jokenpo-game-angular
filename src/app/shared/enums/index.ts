export * from './end-game-result';
export * from './game-status';
export * from './hand-type';
export * from './player-type';
export * from './type-match';
