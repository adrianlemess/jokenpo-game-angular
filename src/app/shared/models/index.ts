export * from './paper-hand.model';
export * from './rock-hand.model';
export * from './scissor-hand.model';
