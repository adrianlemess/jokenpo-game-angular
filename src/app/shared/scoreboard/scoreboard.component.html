<mat-card class="scoreboard">
    <mat-card-header class="scoreboard__header">
        <mat-card-title>{{ scoreboardTitle }}</mat-card-title>
    </mat-card-header>
    <mat-card-content>
        <div class="row">
            <div class="col">
                <mat-card-subtitle>First Player: {{ firstPlayer.username }}</mat-card-subtitle>
            </div>
            <div class="col">
                <mat-card-subtitle>Second Player: {{ secondPlayer.username }}</mat-card-subtitle>
            </div>
        </div>
        <div class="row">
            <div class="col">
                <div
                    class="scoreboard__score"
                    [ngClass]="{
                        'scoreboard__score-winner': firstPlayer.score > secondPlayer.score
                    }"
                >
                    {{ firstPlayer.score }}
                </div>
            </div>
            <div class="col">
                <span class="scoreboard__versus-sign"> X </span>
            </div>
            <div class="col">
                <div
                    class="scoreboard__score"
                    [ngClass]="{
                        'scoreboard__score-winner': secondPlayer.score > firstPlayer.score
                    }"
                >
                    {{ secondPlayer.score }}
                </div>
            </div>
        </div>
    </mat-card-content>
</mat-card>
