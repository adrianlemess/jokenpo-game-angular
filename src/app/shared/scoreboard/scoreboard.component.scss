@import '../../../scss/variables.scss';

.scoreboard {
    text-align: center;
    width: 300px;
    margin: 32px auto;
    padding: 16px 32px;

    &__subtitle {
        text-transform: capitalize;
    }

    &__header {
        justify-content: center;
    }
    &__versus-sign {
        font-size: 48px;
        color: $red;
        margin-top: 24px;
        display: flex;
        justify-content: center;
    }

    &__score {
        box-shadow: 4px 9px 15px $gray_shadow;
        font-size: 3em;
        color: $black;
        font-family: 'Roboto', sans-serif;
        font-weight: 700;
        padding: 13px 20px;
        width: 48px;
        height: 48px;
        display: flex;
        align-items: center;
        justify-content: center;
        background: $gray_card;

        &-winner {
            background: $primary_color;
            color: white;
        }
    }
}
