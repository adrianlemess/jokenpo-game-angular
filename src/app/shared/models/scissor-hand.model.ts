import { END_GAME_RESULT, HAND_TYPE } from '../enums';
import { JokenpoStrategy, Hand } from '../interfaces';

const IMAGE_URL = '';

export class ScissorHand implements Hand {
    name: HAND_TYPE;
    avatarImgUrl: string;
    strategy: JokenpoStrategy;

    constructor(strategy: JokenpoStrategy) {
        this.name = HAND_TYPE.SCISSOR;
        this.avatarImgUrl = IMAGE_URL;
        this.strategy = strategy;
    }

    playAgainst(hand: Hand): END_GAME_RESULT {
        return this.strategy.playAgainst(hand);
    }
}
