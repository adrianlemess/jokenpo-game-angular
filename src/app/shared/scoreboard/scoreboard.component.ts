import { Component, Input, OnInit } from '@angular/core';
import { Player } from '../../core/models';

@Component({
    selector: 'app-scoreboard',
    templateUrl: './scoreboard.component.html',
    styleUrls: ['./scoreboard.component.scss'],
})
export class ScoreboardComponent implements OnInit {
    @Input()
    firstPlayer: Player;

    @Input()
    secondPlayer: Player;

    @Input()
    scoreboardTitle: string;

    constructor() {}

    ngOnInit(): void {}
}
