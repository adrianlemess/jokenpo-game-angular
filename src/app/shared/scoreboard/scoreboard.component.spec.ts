import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MatCardModule } from '@angular/material/card';
import { PLAYER_TYPE } from '../../core/enums';
import { Player } from '../../core/models';

import { ScoreboardComponent } from './scoreboard.component';

describe('ScoreboardComponent', () => {
    let component: ScoreboardComponent;
    let fixture: ComponentFixture<ScoreboardComponent>;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            declarations: [ScoreboardComponent],
            imports: [MatCardModule],
        }).compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(ScoreboardComponent);
        component = fixture.componentInstance;
        const firstPlayer = new Player({
            username: 'Luke Skywalker',
            score: 0,
            playerType: PLAYER_TYPE.HUMAN,
        });

        const secondPlayer = new Player({
            username: 'R2D2',
            score: 0,
            playerType: PLAYER_TYPE.COMPUTER,
        });
        component.scoreboardTitle = 'Test';
        component.firstPlayer = firstPlayer;
        component.secondPlayer = secondPlayer;
        fixture.detectChanges();
    });

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('Should show both with the same color when is tie the score', () => {
        component.firstPlayer.score = 0;
        component.secondPlayer.score = 0;
        fixture.detectChanges();

        expect(fixture).toMatchSnapshot('Tie score');
    });

    it('Should show the first with a different color when the first player is winning', () => {
        component.firstPlayer.score = 1;
        component.secondPlayer.score = 0;
        fixture.detectChanges();

        expect(fixture).toMatchSnapshot('Tie score');
    });

    it('Should show the second with a different color when the second player is winning', () => {
        component.firstPlayer.score = 0;
        component.secondPlayer.score = 2;
        fixture.detectChanges();

        expect(fixture).toMatchSnapshot('Tie score');
    });
});
