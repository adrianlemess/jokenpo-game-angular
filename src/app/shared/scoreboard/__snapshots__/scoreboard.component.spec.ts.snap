// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreboardComponent Should show both with the same color when is tie the score: Tie score 1`] = `
<app-scoreboard
  __ngContext__={[Function LRootView]}
  firstPlayer={[Function Player]}
  scoreboardTitle={[Function String]}
  secondPlayer={[Function Player]}
>
  <mat-card
    class="mat-card mat-focus-indicator scoreboard"
  >
    <mat-card-header
      class="mat-card-header scoreboard__header"
    >
      <div
        class="mat-card-header-text"
      >
        <mat-card-title
          class="mat-card-title"
        >
          Test
        </mat-card-title>
      </div>
    </mat-card-header>
    <mat-card-content
      class="mat-card-content"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            First Player: Luke Skywalker
          </mat-card-subtitle>
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            human
          </mat-card-subtitle>
        </div>
        <div
          class="col"
        >
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            Second Player: R2D2
          </mat-card-subtitle>
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            computer
          </mat-card-subtitle>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="scoreboard__score"
            data-testid="first-player-score"
          >
             0 
          </div>
        </div>
        <div
          class="col"
        >
          <span
            class="scoreboard__versus-sign"
          >
             X 
          </span>
        </div>
        <div
          class="col"
        >
          <div
            class="scoreboard__score"
            data-testid="second-player-score"
          >
             0 
          </div>
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</app-scoreboard>
`;

exports[`ScoreboardComponent Should show the first with a different color when the first player is winning: Win score first player 1`] = `
<app-scoreboard
  __ngContext__={[Function LRootView]}
  firstPlayer={[Function Player]}
  scoreboardTitle={[Function String]}
  secondPlayer={[Function Player]}
>
  <mat-card
    class="mat-card mat-focus-indicator scoreboard"
  >
    <mat-card-header
      class="mat-card-header scoreboard__header"
    >
      <div
        class="mat-card-header-text"
      >
        <mat-card-title
          class="mat-card-title"
        >
          Test
        </mat-card-title>
      </div>
    </mat-card-header>
    <mat-card-content
      class="mat-card-content"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            First Player: Luke Skywalker
          </mat-card-subtitle>
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            human
          </mat-card-subtitle>
        </div>
        <div
          class="col"
        >
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            Second Player: R2D2
          </mat-card-subtitle>
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            computer
          </mat-card-subtitle>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="scoreboard__score scoreboard__score-winner"
            data-testid="first-player-score"
          >
             1 
          </div>
        </div>
        <div
          class="col"
        >
          <span
            class="scoreboard__versus-sign"
          >
             X 
          </span>
        </div>
        <div
          class="col"
        >
          <div
            class="scoreboard__score"
            data-testid="second-player-score"
          >
             0 
          </div>
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</app-scoreboard>
`;

exports[`ScoreboardComponent Should show the second with a different color when the second player is winning: Win score second player 1`] = `
<app-scoreboard
  __ngContext__={[Function LRootView]}
  firstPlayer={[Function Player]}
  scoreboardTitle={[Function String]}
  secondPlayer={[Function Player]}
>
  <mat-card
    class="mat-card mat-focus-indicator scoreboard"
  >
    <mat-card-header
      class="mat-card-header scoreboard__header"
    >
      <div
        class="mat-card-header-text"
      >
        <mat-card-title
          class="mat-card-title"
        >
          Test
        </mat-card-title>
      </div>
    </mat-card-header>
    <mat-card-content
      class="mat-card-content"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            First Player: Luke Skywalker
          </mat-card-subtitle>
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            human
          </mat-card-subtitle>
        </div>
        <div
          class="col"
        >
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            Second Player: R2D2
          </mat-card-subtitle>
          <mat-card-subtitle
            class="mat-card-subtitle scoreboard__subtitle"
          >
            computer
          </mat-card-subtitle>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="scoreboard__score"
            data-testid="first-player-score"
          >
             0 
          </div>
        </div>
        <div
          class="col"
        >
          <span
            class="scoreboard__versus-sign"
          >
             X 
          </span>
        </div>
        <div
          class="col"
        >
          <div
            class="scoreboard__score scoreboard__score-winner"
            data-testid="second-player-score"
          >
             2 
          </div>
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</app-scoreboard>
`;
